File: examples/GenerateData_RigidBody.m
%% GenerateData_RigidBody
% Simulate a sample path for the randomly forced rigid body described by
% the following Ito SDE:
%
%  d(Theta) = LIB(Theta)*omega*dt                                      (1a)
%  d(omega) = inv(I)*(cross(I*omega,omega))*dt + RIB'(Theta)*Sigma*dW  (1b)
%
close all
clear
clc
addpath models
addpath ../src/dynamics

% Repeatable results
rng('default')

% Dimensions
nx = 6;
nw = 3;
nz = 6;

% System parameters
params.I = [  1,   0, 0.1;
              0, 1.5,   0;
            0.1,   0, 0.5];

%% Continuous-time simulation of SDE

% Simulation and sampling times
T = 10;
dt = 0.01;
t = (0:dt:T).';
N = length(t);
dtEM = 1e-4;
NEM = round(T/dtEM) + 1;

% Pre-generate a standard Wiener process so it can be easily saved
W = zeros(NEM,nw);
for k = 2:NEM
    W(k,:) = W(k-1,:) + sqrt(dtEM)*randn(1,nw);
end

% Square root of process noise PSD
Sigma = diag([1 1 0.5]);

% Create an SDE using the example dift and diffusion models
f = @driftModel_RigidBody;
D = @diffusionModel_RigidBody;
SDE = stochasticDifferentialEquation(f,D,Sigma);

% Initial condition
x0 = [zeros(3,1);randn(3,1)];

% Simulate the SDE using the Euler-Maruyama scheme
[x,tSim,xSim,wtil] = SDE.simulate(t,[],x0,dtEM,params,W);

% Measurment noise
R = 0.01*eye(6);
v = (chol(R)*randn(nz,N)).';

% Corrupt measurements with noise
y = zeros(N,nz);
for k = 1:N
    y(k,:) = measurementModel_RigidBody(k,x(k,:).',[],params).';
end
z = y + v;

%% Disrete-time data using a large step size

% Save sample time step
params.dt = dt;

% Simulate using the downsampled Wiener process previously generated
w = interp1(tSim,W,t,"previous");
[~,~,xd] = SDE.simulate(t,[],x0,dt,params,w);

% Discrete-time process noise and covariance
Q = Sigma*Sigma'*dt;

% Corrupt measurements with same noise as the continuous-time case
yd = zeros(N,nz);
for k = 1:N
    yd(k,:) = measurementModel_RigidBody(k,xd(k,:).',[],params).';
end
zd = yd + v;

%% Save and plot results

% Save data
vars = {'t','dt','x','xd','y','z','zd','wtil','W','v','w','Sigma','Q',...
        'R','params','tSim','xSim','SDE'};
save('./data/ExampleData_RigidBody.mat',vars{:})

% Plot states
figure
hold on
plot(tSim,xSim)
plot(t,xd,'--')
hold off

% Plot measurements
figure
hold on
plot(t,y)
plot(t,z)
hold off

File: examples/GenerateData_RigidBody.jl
using LinearAlgebra
import DifferentialEquations as ODE
import StochasticDiffEq as SDE
using JLD2
import MAT
include("./models/rigid_body.jl")
using BenchmarkTools

# Dimensions
const nx = 6
const nw = 3
const nz = 6

# Scale diffusion matrix by Sigma, the square root of the process noise PSD
Sigma = Diagonal([1.0, 1.0, 0.5])

# Measurement noise covariance
R = 0.01*Matrix{Float64}(I,nz,nz)

# Rigid body moment of inertia
Ibody = [1.0 0 0.1;0 1.5 0;0.1 0 0.5]
params = (I = Ibody,)

# Final time
T = 10

# Sample times
dt = 0.01
t = 0:dt:T
N = length(t)

# Euler-Maruyama time step
dt_EM = 1.0e-4

# Drift and diffusion functions from the RigidBody module
function drift!(dx, x, params, t)
    f,_ = RigidBody.driftModel(t, x, nothing, params)
    dx .= f
end
function diffusion!(dG, x, params, t)
    D,_ = RigidBody.diffusionModel(t, x, nothing, params)
    dG .= D*Sigma
end

# Initial condition
x0 = [zeros(3); randn(3)]

# Set up the SDE problem
prob = SDE.SDEProblem(drift!, diffusion!, x0, (0,T), params, noise_rate_prototype = zeros(nx,nw), save_noise = true)

# Solve using the Euler-Maruyama scheme
sol = SDE.solve(prob, SDE.EM(), dt = dt_EM)

# Get states at discrete sample times, t.
x = sol(t).u

# Measurement noise
L = cholesky(R).L
v = [L * randn(nz) for _ in 1:N]  # Vector{Vector{Float64}}, each length nz

# Measurement model
y = [Vector{Float64}(undef, nz) for _ in 1:N]
for k in 1:N
    yk, _ = RigidBody.measurementModel(t[k],x[k],nothing,params)
    y[k] = yk
end

# Add measurement noise
z = y + v

# Save data to a JLD2 file
outfile = joinpath(@__DIR__,"data","JuliaData_RigidBody.jld2")
@save outfile Sigma R params x0 sol t x y z

# Also save compatible data to MAT file
vars = Dict(
    "Sigma" => Sigma,
    "R" => R,
    "I" => Ibody,
    "x0" => x0,
    "t" => hcat(t),
    "x" => hcat(x...),
    "y" => hcat(y...),
    "z" => hcat(z...)
)
outfile = joinpath(@__DIR__,"data","JuliaData_RigidBody.mat")
MAT.matwrite(outfile, vars)
